File: services/graphql-crm/src/mutation/cases.rs
use async_graphql::{Context, InputObject, Object};
use chrono::DateTime;
use chrono::Utc;
use uuid::Uuid;

use crate::models::cases;
use crate::models::enums::{CasePriority, CaseStatus, CaseType};

#[derive(Debug, Clone, InputObject)]
pub struct CreateCaseInput {
    pub case_number: String,
    pub status: Option<CaseStatus>,
    pub priority: Option<CasePriority>,
    pub r#type: Option<CaseType>,
    pub owner_id: Uuid,
    pub contact_id: Option<Uuid>,
    pub description: Option<String>,
}

#[derive(Debug, Clone)]
pub struct Mutation;

#[Object(name = "CrmCasesMutations")]
impl Mutation {
    async fn create_case(
        &self,
        ctx: &Context<'_>,
        payload: CreateCaseInput,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn update_case_number(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        case_number: String,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn update_case_status(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        status: Option<CaseStatus>,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn update_case_priority(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        priority: Option<CasePriority>,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn update_case_type(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        r#type: Option<CaseType>,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn update_case_owner_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        owner_id: Uuid,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn update_case_contact_id(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        contact_id: Option<Uuid>,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn update_case_description(
        &self,
        ctx: &Context<'_>,
        id: Uuid,
        description: Option<String>,
    ) -> async_graphql::Result<cases::Model> {
        todo!()
    }
    async fn remove_case(&self, ctx: &Context<'_>, id: Uuid) -> async_graphql::Result<String> {
        todo!()
    }
}

File: src/graphql/mutations/crm/cases.ts
import { graphql } from "@/lib/graphql/client";

export const createCase = graphql(`
  mutation CreateCase($payload: CreateCaseInput!) {
    crm {
      createCase(payload: $payload) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const updateCaseNumber = graphql(`
  mutation UpdateCaseNumber($id: UUID!, $caseNumber: String!) {
    crm {
      updateCaseNumber(id: $id, caseNumber: $caseNumber) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const updateCaseStatus = graphql(`
  mutation UpdateCaseStatus($id: UUID!, $status: CaseStatus) {
    crm {
      updateCaseStatus(id: $id, status: $status) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const updateCasePriority = graphql(`
  mutation UpdateCasePriority($id: UUID!, $priority: CasePriority) {
    crm {
      updateCasePriority(id: $id, priority: $priority) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const updateCaseType = graphql(`
  mutation UpdateCaseType($id: UUID!, $type: CaseType) {
    crm {
      updateCaseType(id: $id, type: $type) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const updateCaseOwnerId = graphql(`
  mutation UpdateCaseOwnerId($id: UUID!, $ownerId: UUID!) {
    crm {
      updateCaseOwnerId(id: $id, ownerId: $ownerId) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const updateCaseContactId = graphql(`
  mutation UpdateCaseContactId($id: UUID!, $contactId: UUID) {
    crm {
      updateCaseContactId(id: $id, contactId: $contactId) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const updateCaseDescription = graphql(`
  mutation UpdateCaseDescription($id: UUID!, $description: String) {
    crm {
      updateCaseDescription(id: $id, description: $description) {
        id
        caseNumber
        description
        status
        priority
        type
        createdAt
        updatedAt
        owner {
          id
          name
        }
        contact {
          id
          name
        }
      }
    }
  }
`);

export const removeCase = graphql(`
  mutation RemoveCase($id: UUID!) {
    crm {
      removeCase(id: $id)
    }
  }
`);
